pkg/kube/secrets/manifest.go

```diff
@@ -0,0 +1,241 @@
+package secrets
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"code.cloudfoundry.org/cf-operator/pkg/bosh/manifest"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+
+	"k8s.io/client-go/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	manifestKeyName   = "manifest"
+	versionKeyName    = "version"
+	sourceKeyName     = "source-description"
+	deploymentKeyName = "deployment-name"
+)
+
+// ManifestPersister is the interface to persist manifests in Kubernetes
+//
+// It uses Kubernetes secrets to persist a manifest. There can only be one
+// manifest per deployment. Each update to the manifest results in a new
+// persisted version. An existing persisted version of a manifest cannot be
+// altered or deleted. The deletion of a manifest will result in the removal
+// of all persisted version of that manifest.
+//
+// The version number is an integer that is incremented with each version of
+// the manifest, which the greatest number being the current/latest version.
+type ManifestPersister interface {
+	// Creates a new secret version of the provided manifest and a description of the "sources" used to render the manifest (e.g. the location of the CRD that generated it)
+	PersistManifest(manifest manifest.Manifest, sourceDescription string) (*corev1.Secret, error)
+
+	// Deletes all persisted versions of the manifest
+	DeleteManifest() error
+
+	// Adds a label to the current/latest version on the manifest
+	DecorateManifest(key string, value string) (*corev1.Secret, error)
+
+	// Lists all persisted versions of the manifest
+	ListAllVersions() ([]corev1.Secret, error)
+
+	// Get a specific version of the manifest
+	RetrieveVersion(version int) (*corev1.Secret, error)
+
+	// Get the current/latest version of the manifest
+	RetrieveLatestVersion() (*corev1.Secret, error)
+}
+
+type manifestPersisterImpl struct {
+	client         kubernetes.Interface
+	namespace      string
+	deploymentName string
+}
+
+// NewManifestPersister returns a ManifestPersister implementation to be used when working with desired manifest secrets
+func NewManifestPersister(client kubernetes.Interface, namespace string, deploymentName string) ManifestPersister {
```

Comment:
* returns interface
I would have returned the implementation instead of the interface, it still fulfills the contract. Of course that only works if your implementation is public. Your approach has the advantage of providing very tight, Java-ish, encapsulation :)
I remember running into problems testing structs like that in the past, but that might have been me. Anyhow, I enjoyed reading https://mycodesmells.com/post/accept-interfaces-return-struct-in-go
* `uses kubernetes.Interface`
The controllers use a wrapped client from `controller-runtime`, it's not too different. I mostly use it, because I'm under the impression it should provide caching. Not sure where this feature will be used and what client it should use.
* returning a pointer
As far as I understand it's beneficial to return a pointer if the struct uses a pointer receiver. However I'm not sure the `ManifestPersisterImpl` should use a pointer receiver :)  It doesn't look like any of the fields are being mutated, so it could get by with a regular receiver?

Reply:
I changed the code with regards to the comments.